### ISSUE TITLE
Is it possible to support parallel NetCDF I/O?

### ISSUE THREAD

**Issue by ali-ramadhan** [Alexander-Barth/NCDatasets.jl#122]
I don't know much about the subject but from looking at the PnetCDF description (https://parallel-netcdf.github.io/) it sounds like there are two backend options for parallel I/O: PnetCDF and parallel HDF5?

It sounds like it might be possible to build NetCDF with parallel I/O support.

Out of curiousity, is parallel I/O something that NCDatasets.jl can feasibly support?

X-Ref: https://github.com/CliMA/ClimateMachine.jl/issues/2007

**Comment by Alexander-Barth** [comment]
Support for parallel NetCDF I/O would indeed be nice.
 Do you have any complete known-working C code example?
I tried the Ubuntu 20.04 package libnetcdf-mpi-dev with [nc4_pnc_put.c](http://cucis.ece.northwestern.edu/projects/PnetCDF/Examples/nc4_pnc_put.c)

But this fails with:
```
sudo apt-get install libnetcdf-mpi-dev
wget http://cucis.ece.northwestern.edu/projects/PnetCDF/Examples/nc4_pnc_put.c
gcc -o nc4_pnc_put -I/usr/lib/x86_64-linux-gnu/netcdf/mpi/include/ nc4_pnc_put.c -L/usr/lib/x86_64-linux-gnu/netcdf/mpi/ -lnetcdf_mpi -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -pthread -L/usr/lib/x86_64-linux-gnu/openmpi/lib -lmpi
mpiexec -n 2 ./nc4_pnc_put testfile.nc
# Error at line=100: NetCDF: Parallel operation on file opened for non-parallel access Aborting ...
```
It could be that I did something stupid.
I assume that the two package have the same API. Maybe this is not the case.

If don't know much on this subject either, but I somebody could provide a full C example (as opposed to code fragments) that would help a lot.

**Comment by ali-ramadhan** [comment]
Thanks for looking into this. I thought I was going to start playing around with parallel I/O sooner but still working on basic MPI infrastructure...

I can try getting `nc4_pnc_put.c` to run once I start looking at parallel NetCDF I/O.

**Comment by rafaqz** [comment]
Any luck with this @ali-ramadhan? netcdf could also be a sink for parallel DiskArrays/Dagger.jl processing, so this would be widely useful.

**Comment by kongdd** [comment]
call for this feature too

**Comment by johnomotani** [comment]
:eyes:

**Comment by rafaqz** [comment]
JLL packages are not so hard to add with the wizard (see https://github.com/JuliaPackaging/BinaryBuilder.jl)

And it is _very_ likely that you are the best situated person to do this currently, probably the 100 to 1 favorite.

And code that relies on manually installed system binaries will likely not be widely used. The julia ecosystem has moved very strongly towards completely versions dependencies managed by Pkg.jl. 

So, I strongly encourage you to give it a go :) 

But feel free to ping any JLL problems here for help/feedback.

**Comment by johnomotani** [comment]
> And code that relies on manually installed system binaries will likely not be widely used. The julia ecosystem has moved very strongly towards versioned dependencies managed by Pkg.jl.

Just to chime in here (as a potential parallel NetCDF user!) - parallel NetCDF (or HDF5) is one case where system binaries _are_ likely to be wanted. On an HPC cluster, we probably have to use the vendor-provided MPI to get the best performance (especially for inter-node communication), so the parallel NetCDF (and HDF5, which I mention as it'll be a dependency of parallel NetCDF for netcdf4 files, I assume) libraries will need to be linked to the system MPI, which the Julia-provided binaries will not be. At least, HPC users will want the option to do that, and I guess they are the main users of parallel NetCDF...

For comparison, see the setup for parallel HDF.jl, which provides a utility function to link to the system binaries:
https://juliaio.github.io/HDF5.jl/stable/mpi/#using_parallel_HDF5

**Comment by rafaqz** [comment]
Yes you're probably right generally, I was thinking of non-MPI use cases like Dagger.jl.

This is pretty nice syntax in HDF5 if we only have the system binaries:

`HDF5.API.set_libraries!("/path/to/your/libhdf5.so", "/path/to/your/libhdf5_hl.so")`

**Comment by Alexander-Barth** [comment]
There is some initial work in https://github.com/Alexander-Barth/NCDatasets.jl/commit/70ef6830497674461c33114940ae58d439c7f827 in the branch MPI.
For now I prioritize the current HDF5/NetCDF4 format and the Yggdrasil JLLs.

**Comment by Alexander-Barth** [comment]
Windows currently fails with ([full logs](https://github.com/Alexander-Barth/NCDatasets.jl/actions/runs/8806666199/job/24171986212)):

```
NetCDF: Parallel operation on file opened for non-parallel access (NetCDF error code: -114 = NC_ENOPAR)
   nc_create_par(path::String, cmode::UInt16, mpi_comm::MPI.Comm, mpi_info::MPI.Info)
   NCDataset(comm::MPI.Comm, filename::String, mode::String; info::MPI.Info, format::Symbol, share::Bool, diskless::Bool, persist::Bool, maskingvalue::Missing, attrib::Vector{Any})
  in expression starting at D:\a\NCDatasets.jl\NCDatasets.jl\test\test_mpi_script.jl:18
  in expression starting at D:\a\NCDatasets.jl\NCDatasets.jl\test\test_mpi_netcdf.jl:10
```

Parallel support seem to be missing from the Windows NetCDF_jll
https://github.com/JuliaBinaryWrappers/NetCDF_jll.jl/releases/download/NetCDF-v400.902.211%2B0/NetCDF-logs.v400.902.211.x86_64-w64-mingw32-mpi+microsoftmpi.tar.gz


```
checking whether parallel io is enabled in hdf5... no
checking for library containing H5Dread_chunk... none required
checking for library containing H5Pset_fapl_ros3... none required
checking whether HDF5 allows parallel filters... yes
checking whether szlib was used when building HDF5... yes
checking whether HDF5 library is version 1.10.6 or later... yes
configure: WARNING: Parallel io disabled for netcdf-4 because hdf5 does not support
checking whether parallel I/O is enabled for netcdf-4... no
```


```
# Features
--------
Benchmarks:		no
NetCDF-2 API:		yes
HDF4 Support:		no
HDF5 Support:		yes
NetCDF-4 API:		yes
CDF5 Support:		yes
NC-4 Parallel Support:	no
PnetCDF Support:	no
```

**Comment by Alexander-Barth** [comment]
It is not clear if HDF5 has actually MPI enabled on Windows:

https://github.com/JuliaBinaryWrappers/HDF5_jll.jl/releases/download/HDF5-v1.14.3%2B3/HDF5-logs.v1.14.3.x86_64-w64-mingw32-libgfortran3-cxx03-mpi+microsoftmpi.tar.gz

```
Features:
---------
                     Parallel HDF5: no
  Parallel Filtered Dataset Writes: no
                Large Parallel I/O: no
```

If somebody with an interesting in Windows can have a look at this this would be awesome :-).

**Comment by pgf** [comment]
> There is some initial work in [70ef683](https://github.com/Alexander-Barth/NCDatasets.jl/commit/70ef6830497674461c33114940ae58d439c7f827) in the branch MPI.

Is it new? I don't remember seeing it when I checked last week.
Anyway, it's almost identical to my version, except for few details. For example I called the `access` method `paraccess` to be more explicit, but that's fine.

A couple of notes:

- the `access` dataset method works only with netcdf3 files, while it throws an error with netcdf4 files because `nc_var_par_access` doesn't recognize `NC_GLOBAL` as a valid variable ID. AFAIK when `nc_var_par_access` is called on a variable in a netcdf3 file it sets the access mode globally for the file (see [here](https://github.com/Unidata/netcdf-c/blob/fc02ef4e75b97392f6361b286ff3bf60b013de1c/libsrcp/ncpdispatch.c#L1153)). I think there's no need for a dataset method, the variable method already does the same with netcdf3 files. Indeed, I too added the dataset method at first, but later I changed my mind.
- I think that the MPI communicator can be an optional argument for most of the cases, so the user just need to ask for parallel access. I defined the `NCDataset` method this way:

   ```julia
   function NCDataset(filename::AbstractString,
                   mode::AbstractString = "r";
                   format::Symbol = :netcdf4,
                   parallel::Bool = false,
                   comm::MPI.Comm = MPI.COMM_WORLD,
                   info::MPI.Info = MPI.INFO_NULL,
                   ...
   ```
   I think this gives a cleaner call (similar to the python [netcdf4](https://unidata.github.io/netcdf4-python/#netCDF4.Dataset) package):
   ```julia
   ds = NCDataset(path,"c",parallel=true)
   ```
   leaving `comm` and `info` for more specific use cases.
